src/Vacuum/Vacuum_math.jl

```diff
@@ -0,0 +1,728 @@
+- `ax::AbstractVector{Float64}`: Array of x-coordinates for the interpolation points.
+- `af::AbstractVector{Float64}`: Array of y-coordinates (function values) for the interpolation points.
+- `m::Int`: Number of interpolation points.
+- `nl::Int`: Number of points to use for the local interpolation (degree of polynomial + 1).
+- `x::Float64`: The x-value at which to evaluate the interpolated function and/or its derivative.
+
+# Returns
+- `f::Float64`: The interpolated function value at `x`
+- `df::Float64`: The interpolated function derivative at `x` 
+"""
+function lagrange1d!(ax::Vector{Float64}, af::Vector{Float64}, m::Int, nl::Int, x::Float64, f::Float64, df::Float64)
+
+    jn = findfirst(i -> ax[i] >= x, 1:m)
+    jn = (jn === nothing) ? m : jn
+    jn = max(jn - 1, 1)
+    if jn < m && abs(ax[jn + 1] - x) < abs(x - ax[jn])
+        jn += 1
+    end
+
+    # Determine the range of indices for interpolation
+    jnmm = floor(Int, (nl - 0.1) / 2)
+    jnpp = floor(Int, (nl + 0.1) / 2)
+
+    nll = jn - jnmm
+    nlr = jn + jnpp
+
+    # Adjust for even nl when ax[jn] > x
+    if (nl % 2 == 0) && (ax[jn] > x)
+        nll -= 1
+        nlr += 1
+    end
+
+    # Clamp indices to valid array bounds
+    if nlr > m
+        nlr = m
+        nll = nlr - nl + 1
+    elseif nll < 1
+        nll = 1
+        nlr = nl
+    end
+
+    # Compute function value f
+    for i in nll:nlr
+        alag = 1.0
+        for j in nll:nlr
+            if i == j
+                continue
+            end
+            alag *= (x - ax[j]) / (ax[i] - ax[j])
+        end
+        f += alag * af[i]
+    end
+    if iop == 0
+        return f, df # df is 0.0 as initialized
+    end
+
+    # Compute derivative df
+    for i in nll:nlr
+        slag = 0.0
+        for id in nll:nlr
+            if id == i
+                continue
+            end
+            alag = 1.0
+            for j in nll:nlr
+                if j == i
+                    continue
+                end
+                if j != id
+                    alag *= (x - ax[j]) / (ax[i] - ax[j])
+                else
+                    alag /= (ax[i] - ax[id])
+                end
+            end
+            slag += alag
+        end
+        df += slag * af[i]
+    end
+end
+
+#############################################################
+# smoothing array, replacing smooth0, smooth
+#############################################################
+function smooth(y::Vector{Float64}, w::Int)
+    n = length(y)
+    y_smooth = similar(y)
+    for i in 1:n
+        i1 = max(1, i - div(w,2))
+        i2 = min(n, i + div(w,2))
+        y_smooth[i] = mean(y[i1:i2])
+    end
+    return y_smooth
+end
+
+#############################################################
+# shift array, replacing shift
+#############################################################
+function shift(y::Vector, nshift::Int)
+    n = length(y)
+    nshift = mod(nshift, n)
+    return vcat(y[end-nshift+1:end], y[1:end-nshift])
```

Comment:
Again, avoid slicing with `@views`. I also think [`circshift`](https://docs.julialang.org/en/v1/base/arrays/#Base.circshift) does what you want.